lms/djangoapps/bookmarks/tests/test_views.py

```diff
@@ -0,0 +1,462 @@
+
+    def test_get_bookmarks_with_pagination(self):
+        """
+        Test that requesting bookmarks for a course return results with pagination 200 code.
+        """
+        query_parameters = 'course_id={}&page_size=1'.format(self.course_id)
+        response = self.send_get(client=self.client, url=reverse('bookmarks'), query_parameters=query_parameters)
+
+        bookmarks = response.data['results']
+
+        # Pagination assertions.
+        self.assertEqual(response.data['count'], 2)
+        self.assertIn('page=2&page_size=1', response.data['next'])
+        self.assertEqual(response.data['num_pages'], 2)
+
+        self.assertEqual(len(bookmarks), 1)
+        self.assert_valid_bookmark_response(bookmarks[0], self.bookmark_2)
+
+    def test_get_bookmarks_with_invalid_data(self):
+        """
+        Test that requesting bookmarks with invalid data returns a 0 records.
+        """
+        # Invalid course id.
+        response = self.send_get(client=self.client, url=reverse('bookmarks'), query_parameters='course_id=invalid')
+        bookmarks = response.data['results']
+        self.assertEqual(len(bookmarks), 0)
+
+    def test_get_all_bookmarks_when_course_id_not_give(self):
+        """
+        Test that requesting bookmarks returns all records for that user.
+        """
+        # Without course id we would return all the bookmarks for that user.
+        response = self.send_get(client=self.client, url=reverse('bookmarks'))
+        bookmarks = response.data['results']
+        self.assertEqual(len(bookmarks), 2)
+        self.assert_valid_bookmark_response(bookmarks[0], self.bookmark_2)
+        self.assert_valid_bookmark_response(bookmarks[1], self.bookmark_1)
+
+    def test_anonymous_access(self):
+        """
+        Test that an anonymous client (not logged in) cannot call GET or POST.
+        """
+        query_parameters = 'course_id={}'.format(self.course_id)
+        self.send_get(
+            client=self.anonymous_client,
+            url=reverse('bookmarks'),
+            query_parameters=query_parameters,
+            expected_status=401
+        )
+        self.send_post(
+            client=self.anonymous_client,
+            url=reverse('bookmarks'),
+            json_data={"usage_id": "test"},
+            expected_status=401
+        )
+
+    def test_post_bookmark_successfully(self):
+        """
+        Test that posting a bookmark successfully returns newly created data with 201 code.
+        """
+        response = self.send_post(
+            client=self.client,
+            url=reverse('bookmarks'),
+            json_data={'usage_id': unicode(self.vertical_3.location)}
+        )
+
+        # Assert Newly created bookmark.
+        self.assertEqual(response.data['id'], "%s,%s" % (self.user.username, unicode(self.vertical_3.location)))
+        self.assertEqual(response.data['course_id'], self.course_id)
+        self.assertEqual(response.data['usage_id'], unicode(self.vertical_3.location))
```

Comment:
If you decide that always returning OPTIONAL_FIELDS is correct, you should test for them here.